### PR TITLE
max_instances for DatasetReader

### PR DESCRIPTION
**Pull request by dirkgr**
This is from a comment on the TransformerQA PR. Rather than having a one-off limitation on reader activity, this gives the ability to all the readers.

I had to change how the dataset reader tests are run to make `@pytest.mark.parametrize` work.

### CODE REVIEW

File: allennlp/tests/data/dataset_readers/dataset_reader_test.py [hunk 1]
```diff
@@ -1,30 +1,31 @@
...
-        super().setUp()
-        self.cache_directory = str(self.FIXTURES_ROOT / "data_cache" / "with_prefix")
+class TestDatasetReader:
+    cache_directory = str(AllenNlpTestCase.FIXTURES_ROOT / "data_cache" / "with_prefix")
 
-    def tearDown(self):
-        super().tearDown()
-        shutil.rmtree(self.cache_directory)
+    @pytest.yield_fixture(autouse=True)
+    def cache_directory_fixture(self):
```
Comment (DeNeutoy) [inline allennlp/tests/data/dataset_readers/dataset_reader_test.py]: https://docs.pytest.org/en/latest/yieldfixture.html

FYI, looks like this is deprecated for pytest > 3.0 (I have 4.3.1 installed) and you can just `yield` from a `pytest.fixture()` instead.

File: allennlp/data/dataset_readers/dataset_reader.py [hunk 2]
```diff
@@ -84,8 +85,14 @@ class DatasetReader(Registrable):
         parameters you set for this DatasetReader!_
     """
 
-    def __init__(self, lazy: bool = False, cache_directory: str = None) -> None:
+    def __init__(
+        self,
+        lazy: bool = False,
+        cache_directory: Optional[str] = None,
+        max_instances: Optional[int] = None,
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/dataset_reader.py]: Can you add this to the docstring?

File: allennlp/tests/data/dataset_readers/dataset_reader_test.py (hunk 1 above)
Comment (dirkgr) [reply allennlp/tests/data/dataset_readers/dataset_reader_test.py]: fa7b32498974f04bf841f3185b6c9fdf34799990

File: allennlp/data/dataset_readers/dataset_reader.py (hunk 2 above)
Comment (dirkgr) [reply allennlp/data/dataset_readers/dataset_reader.py]: bb753f02638cabaf03a207272d49647179fd6ba4

**Pull request merged**